include/alpaka/mem/view/Accessor.hpp

```diff
@@ -0,0 +1,442 @@
+/* Copyright 2021 Bernhard Manfred Gruber
+
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED “AS IS” AND ISC DISCLAIMS ALL WARRANTIES WITH
+ * REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL ISC BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
+ * IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#pragma once
+
+#include <alpaka/dim/Traits.hpp>
+#include <alpaka/mem/view/Traits.hpp>
+
+#include <tuple>
+
+namespace alpaka
+{
+    namespace internal
+    {
+        template<typename T>
+        ALPAKA_FN_HOST_ACC auto asBytePtr(T* p)
+        {
+            return reinterpret_cast<char*>(p);
+        }
+        template<typename T>
+        ALPAKA_FN_HOST_ACC auto asBytePtr(const T* p)
+        {
+            return reinterpret_cast<const char*>(p);
+        }
+    } // namespace internal
+
+    //! Access tag type indicating read-only access.
+    struct ReadAccess
+    {
+    };
+
+    //! Access tag type indicating write-only access.
+    struct WriteAccess
+    {
+    };
+
+    //! Access tag type indicating read-write access.
+    struct ReadWriteAccess
+    {
+    };
+
+    //! An accessor is an abstraction for accessing memory objects such as views and buffers.
+    //! @tparam TMemoryHandle A handle to a memory object.
+    //! @tparam TElem The type of the element stored by the memory object. Values and references to this type are
+    //! returned on access.
+    //! @tparam TBufferIdx The integral type used for indexing and index computations.
+    //! @tparam TDim The dimensionality of the accessed data.
+    //! @tparam TAccessModes Either a single access tag type or a `std::tuple` containing multiple access tag types.
```

Comment:
I still don't see the use case for having multiple access modes for the same accessor. @psychocoderHPC I heard this was your wish. Could you explain a bit why this is required?

Reply:
Some discussion was in this PR some comments ago: https://github.com/alpaka-group/alpaka/pull/1249#discussion_r575296107

Reply:
But that is  for the "old" design with cursors and accessors, right? Is this still required for accessors without cursors?

Reply:
I think the principle still holds, except that instead of cursors, you will just get another accessor when constraining an accessor.